### PR TITLE
Split Docker images in two: one for base and one for edge.

### PR DESCRIPTION
**Pull request by Zimmi48**
This is an alternative to #17780 applying @SkySkimmer's suggestion of splitting the Docker image in two: one for the base, one for the edge. This should also speed up builds a little, by pulling smaller images.

The base image will use the old Ubuntu LTS (currently 20.04) and the edge image will use the most recent Ubuntu version (currently 23.04).

### CODE REVIEW

**Comment by ejgallego** [comment]
Cool!

Note that it is possible to generate two images from a single docker file, just in case you'd find that more convenient, but I don't think there is a lot of sharing going on here?

**Comment by Zimmi48** [comment]
Todo: I should update the documented  minimal bounds of dependencies to match what is actually going to be tested.

**Comment by Zimmi48** [comment]
@coqbot run full CI

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -31,15 +31,16 @@ stages:
...
-  # Format: bionic_coq-V$DATE-$hash
+  # Format: image_name-V$DATE-$hash
   # $DATE is so we can tell what's what in the image list
   # The $hash is the first 10 characters of the md5 of the Dockerfile. e.g.
-  # echo $(md5sum dev/ci/docker/bionic_coq/Dockerfile | head -c 10)
-  CACHEKEY: "bionic_coq-V2023-06-29-2b9dca3152"
-  IMAGE: "$CI_REGISTRY_IMAGE:$CACHEKEY"
+  # echo $(md5sum dev/ci/docker/old_ubuntu_lts/Dockerfile | head -c 10)
+  # echo $(md5sum dev/ci/docker/edge_ubuntu/Dockerfile | head -c 10)
+  BASE_CACHEKEY: "old_ubuntu_lts-V2023-06-30-42bb810307"
```
Comment (ejgallego) [inline .gitlab-ci.yml:39]: we use `base/edge` terminology on this file, maybe we could imitate that here?

Comment (Zimmi48) [reply .gitlab-ci.yml:39]: I know, but the "old LTS" terminology is on purpose, because it defines how the version of this Ubuntu image should be updated (always use the old LTS version, i.e., the one before the last LTS version).

**Comment by Zimmi48** [comment]
@coqbot run full CI

**Comment by Zimmi48** [comment]
From my testing, this PR should be finished and working now, but let's run a full CI to be sure:

@coqbot run full CI

File: doc/README.md [hunk 2]
```diff
@@ -32,9 +32,9 @@ reference manual requires Python 3, and the following Python packages:
 
   - sphinx >= 4.5.0
   - sphinx_rtd_theme >= 1.0.0
-  - beautifulsoup4 >= 4.0.6
+  - beautifulsoup4 >= 4.8.2
   - antlr4-python3-runtime >= 4.7.1
-  - pexpect >= 4.2.1
+  - pexpect >= 4.6.0
```
Comment (SkySkimmer) [inline doc/README.md:37]: I guess those are the versions in ubuntu 20.04?

**Review by SkySkimmer — APPROVED** [review]

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -310,14 +301,13 @@ build:base+32bit:
...
   variables:
-    OPAM_SWITCH: edge
     OPAM_VARIANT: "+flambda"
     COQ_EXTRA_CONF: "-native-compiler yes"
-  only: *full-ci
 
-build:edge+flambda:dev:
-  extends: .build-template:dev
+build:base:dev:
+  extends: .build-template:base:dev
```
Comment (ejgallego) [inline .gitlab-ci.yml:310]: Why this change?

**Review by ejgallego — CHANGES_REQUESTED** [review]
This PR implements changes to the actual CI process that are not documented in the commit messages / PRs , I think they should be documented.

**Comment by ejgallego** [comment]
My understanding after reading the changes in CI setup is that:
- I think it is a good idea to introduce a "doc" template, however note that the 3 doc jobs have really different envs so I'm unsure if this can work for us. One is a pure ML dev + odoc (for the ML API), the other is a sphinx one, and the other is a coqdoc one.
- I would keep (and maybe document more) the `.ci-dev` template. The point of this template is to pass to subjobs the whole Dune build database so incremental builds are possible (actually for docs I think this is only needed for the ML API one) , but indeed this template maybe prove useful in the future if we do something like #16201 . Also maybe we should switch the sphinx job to use the regular edge template (and fix #14269)
- I think the `:dev` job makes more sense with the egde version of the compiler, due to warnings etc... and because indeed that version is the one we usually recommend to devs

File: doc/README.md (hunk 2 above)
Comment (Zimmi48) [reply doc/README.md:37]: Yes

**Comment by Zimmi48** [comment]
>I think it is a good idea to introduce a "doc" template, however note that the 3 doc jobs have really different envs so I'm unsure if this can work for us. One is a pure ML dev + odoc (for the ML API), the other is a sphinx one, and the other is a coqdoc one.

I didn't introduce a doc template, I just renamed a template that happened to be used exactly by the doc jobs and no other ones.

>Also maybe we should switch the sphinx job to use the regular edge template

No, documentation should be built with the base image, not the edge one, so that we test our minimal documented dependencies.

**Comment by ejgallego** [comment]
> I didn't introduce a doc template, I just renamed a template that happened to be used exactly by the doc jobs and no other ones.

Yes, by "introduce" I really meant "renamed".

> > Also maybe we should switch the sphinx job to use the regular edge template
> 
> No, documentation should be built with the base image, not the edge one, so that we test our minimal documented dependencies.

Yup, base should be fine. I didn't mean base vs edge but `dev` vs `install` setup.

It seems fine to me that the docs are tested with base, however I think the base `dev` build should still happen with edge, that is the change I'm not sure we want here.

So I guess the options I see here are:
- keep the `edge:dev` build, and maybe build the API docs with it?
- introduce a `base:dev` build, and switch the refman + coqdoc builds to it, however, due to #14269 could make sense to just switch those to the regular (base) build

The idea I had was to use the opam jobs to test the doc build with the regular, release, installed build, which should be possible once the coq-doc PR is in.

But indeed we may do differently, as we need to test two things:
- doc in release mode (for upload)
- docs in dev build (to check devs can build docs locally)

So we could have the dev build be a two phase build, and have the first one do the quick, non-release mode OCaml build, and the second, a build that would test the rest of the tooling.

**Comment by Zimmi48** [comment]
I'm fine with all what you propose, but I would also like to avoid the work following on me just because I have taken the task to update our Docker image, and to follow the suggestion that was made then to split the Docker image in two.

Right now, the doc does not yet build outside developer mode, so I propose to merge this PR as is (I can update the commit messages to be more precise, once we agree on what exactly they should say) and to move back to a dev+edge job once the doc can be moved to depend on the release+base one (but I would leave this to you). I think that it is acceptable to have a transitional state where the dev setup is tested with the base image rather than the edge one (IMHO, the dev setup is largely tested anyway by the variety of configurations we all have).

**Comment by Zimmi48** [comment]
FTR the fiat_crypto_ocaml and coqtail failures are probably related because they are about Python. In particular, fiat_crypto_ocaml cannot find `python` (I guess only `python3` exists).

**Comment by ejgallego** [comment]
> I'm fine with all what you propose, but I would also like to avoid the work falling on me just because I have taken the task to update our Docker image and to follow the suggestion that was made then to split the Docker image in two.

I agree, maybe the best is to postpone the change to the `:dev` build to a different PR ?

> Right now, the doc does not yet build outside developer mode

Do you mean that the doc will fail to build or that we haven't added the CI mode?

> so I propose to merge this PR as is (I can update the commit messages to be more precise, once we agree on what exactly they should say) and to move back to a dev+edge job once the doc can be moved to depend on the release+base one (but I would leave this to you). I think that it is acceptable to have a transitional state where the dev setup is tested with the base image rather than the edge one (IMHO, the dev setup is largely tested anyway by the variety of configurations we all have).

I personally think that it would be best to separate the Docker split changes and CI policy changes in two different PRs, but I'm happy if you want to go ahead with the PR as is (and indeed update the description commit message).

**Comment by ejgallego** [comment]
> FTR the fiat_crypto_ocaml and coqtail failures are probably related because they are about Python. In particular, fiat_crypto_ocaml cannot find `python` (I guess only `python3` exists).

A quick fix on ubuntu is to install the `python-is-python3` package.

**Comment by ejgallego** [comment]
> > FTR the fiat_crypto_ocaml and coqtail failures are probably related because they are about Python. In particular, fiat_crypto_ocaml cannot find `python` (I guess only `python3` exists).
> 
> A quick fix on ubuntu is to install the `python-is-python3` package.

But we should open an issue so we fix this at some point indeed, but I think that indeed you don't want to tackle that problem in this PR

**Comment by Zimmi48** [comment]
> I agree, maybe the best is to postpone the change to the `:dev` build to a different PR ?

That would mean that we would stop testing the doc with the base image, which is IMO more problematic than to stop testing the dev build with the edge image.

> Do you mean that the doc will fail to build or that we haven't added the CI job to do so?

I haven't tested it recently, so I may be wrong, but my recollection is that it would fail.

>I personally think that it would be best to separate the Docker split changes and CI policy changes in two different PRs, but I'm happy if you want to go ahead with the PR as is (and indeed update the description commit message).

This indeed has my preference. Note that this doesn't change anything that is part of the documented CI policy. This does change how things are tested in practice, though.

**Comment by ejgallego** [comment]
> That would mean that we would stop testing the doc with the base image, which is IMO more problematic than to stop testing the dev build with the edge image.

I see what you mean now, ok then, but we using the base compiler for the `dev` build may create problems due to warnings etc... so we should be careful not to forget to update that. Both scenarios are IMO bad.

Actually I was under the impression that for the doc build the image didn't matter a lot, but indeed it does, for example the `pip` command line doesn't work for python 3.10 which is in edge.

> I haven't tested it recently, so I may be wrong, but my recollection is that it would fail.

I was testing recently and things do work fine, only case that is broken is a bit of a corner one (joint release build for a global install #16198 )

> This indeed has my preference. Note that this doesn't change anything that is part of the documented CI policy. This does change how things are tested in practice, though.

Yes, this changes the version of the compiler `:dev` is build with, and the version of `odoc` that is used for the ML API, on top of the expected changes to python and system libs versions.

**Comment by ejgallego** [comment]
> > > FTR the fiat_crypto_ocaml and coqtail failures are probably related because they are about Python. In particular, fiat_crypto_ocaml cannot find `python` (I guess only `python3` exists).
> > 
> > 
> > A quick fix on ubuntu is to install the `python-is-python3` package.
> 
> But we should open an issue so we fix this at some point indeed, but I think that indeed you don't want to tackle that problem in this PR

We should open issues actually not in the Coq repos (which is python3 -free since #10491 )

**Comment by Zimmi48** [comment]
> I see what you mean now, ok then, but we using the base compiler for the `dev` build may create problems due to warnings etc... so we should be careful not to forget to update that. Both scenarios are IMO bad.

Right, I forgot about warnings.

> I was testing recently and things do work fine, only case that is broken is a bit of a corner one (joint release build for a global install #16198 )

OK, in that case, I can try to have the doc build with the install base build, but should I change the commands that are run then?

**Comment by ejgallego** [comment]
> Right, I forgot about warnings.
> 
> > I was testing recently and things do work fine, only case that is broken is a bit of a corner one (joint release build for a global install #16198 )
> 
> OK, in that case, I can try to have the doc build with the install base build, but should I change the commands that are run then?

Yes, but indeed I suggest we do that in different PRs, now that we have clarified the issues with the split, I suggest you finish this PR in the way most comfortable for your and then we do:

- one PR to move the dev build back to edge compiler
- one PR to move the doc jobs to release mode and fix the native + doc bug
- one PR to add a second dev build for all the tooling, docs, etc... that build could use a venv which is more convinent for devs, and use a more modern python package chain

Then the fast CI mode would instead run the two dev builds and leave the doc deployment to the full ci phase.

Only point I see still pending would be coqbot, but indeed that can be discussed on each of the PRs.

WDYT?

**Comment by Zimmi48** [comment]
Sounds good to me!

**Review by ejgallego — APPROVED** [review]

**Comment by JasonGross** [comment]
> If these contribs depend on python2 we are gonna have hard time supporting them as Python2 has been completely removed from Debian.

AFAIR, we don't depend on Python2, but we do depend on having `python` in path (regardless of whether it's 2 or 3).  (Or maybe the issue is that we depend on having `python3` in path?)



> fiat-crypto is fixed to support glibc 2.37

The issue is not with Fiat Crypto but that we use different images on CI for Fiat Crypto vs Fiat Crypto OCaml.  IIRC, this was because native compiler didn't play nice with flambda, but the extracted OCaml files were sad without it?  I don't quite recall.  Anyway, if you can upgrade all the images to use the same OS, then this problem will go away

**Comment by Zimmi48** [comment]
> AFAIR, we don't depend on Python2, but we do depend on having `python` in path (regardless of whether it's 2 or 3). (Or maybe the issue is that we depend on having `python3` in path?)

I think the issue was indeed that only `python3` was provided by Ubuntu by default, not `python`. Personally, I think this is stupid to have removed the `python` alias by default, but well..., at least there is a package that can easily be installed to get it back.

> The issue is not with Fiat Crypto but that we use different images on CI for Fiat Crypto vs Fiat Crypto OCaml. IIRC, this was because native compiler didn't play nice with flambda, but the extracted OCaml files were sad without it? I don't quite recall. Anyway, if you can upgrade all the images to use the same OS, then this problem will go away

Oh, thanks for the explanation! I didn't recall this oddity. Then, there are three options:

- do not build fiat-crypto and related jobs with flambda (that's probably not such a good idea because of the increased build time),
- add a flambda switch to the base image and always use the base image for fiat-crypto related jobs
- add a non-flambda switch to the edge image and always use the edge image for fiat-crypto related jobs

My preference is the third option as it creates the least change. Furthermore, the base image already contains several switches whereas the edge one only contained one so far.

@SkySkimmer @ejgallego any opinion?

**Comment by SkySkimmer** [comment]
>IIRC, this was because native compiler didn't play nice with flambda, but the extracted OCaml files were sad without it?

That's unlikely as the main build is with flambda and the ocaml files are built without flambda.

I think we just used flambda for speed and to have more testing of it.

**Comment by Zimmi48** [comment]
The code says:

```yaml
# We cannot use flambda due to
# https://github.com/ocaml/ocaml/issues/7842, see
# https://github.com/coq/coq/pull/11916#issuecomment-609977375
```

**Comment by SkySkimmer** [comment]
See also https://github.com/coq/coq/pull/12431

**Comment by SkySkimmer** [comment]
>The code says:

That's on the _ocaml job

**Comment by SkySkimmer** [comment]
>Oh, thanks for the explanation! I didn't recall this oddity. Then, there are three options:

It might also work to change the `needs` of the fiat_crypto_ocaml job to use build:base instead of edge build

TBH changing switches / images has felt hacky to me from the start so I lean towards not using flambda or dropping the _ocaml job.

**Comment by ejgallego** [comment]
> Personally, I think this is stupid to have removed the python alias by default

This was done to anticipate `python4`, the mess that `python` became Python3 from Python2 was epic I heard.

Regarding the Fiat Crypto issue, I'm surprised it doesn't work in flambda. I'd try to put everything in edge and see what happens.

If the compiler overflows, `ulimit` should take care of it, I was never able to make it stack overflow with a 32MiB stack.

**Comment by ejgallego** [comment]
@JasonGross the issue is that you use `python` which doesn't exist anymore in newer Ubuntu / Debian.

Python policy requires scripts to use the versioned form of the binary.

**Comment by Zimmi48** [comment]
Based on the answers, I will try first the version where fiat_crypto_ocaml is moved to the edge+flambda build, then the version where all of its dependencies move to base, if the previous strategy did not work.

**Comment by Zimmi48** [comment]
@ejgallego 

> If the compiler overflows, `ulimit` should take care of it, I was never able to make it stack overflow with a 32MiB stack.

In https://github.com/ocaml/ocaml/issues/7842#issuecomment-611200117 you said:

>This is becoming more and more common, for example https://github.com/coq/coq/pull/12036 , even worse, for some files not even increasing the stack size will work [unless ridicoulous values are used]

At the time, did you consider a 32MiB stack ridiculous values?

I'm trying to determine which exact `ulimit` call I should make in the `fiat_crypto_ocaml` config.

**Comment by Zimmi48** [comment]
Actually, `dev/ci/ci-fiat_crypto.sh` already sets the stack size to 32 MB. And it does so with a comment linking to https://github.com/coq/coq/pull/8313#issuecomment-416650241 (so a comment from 2018), whereas the fiat_crypto_ocaml issue was more recent, so I'm thinking we won't be able to solve it by increasing the stack size.

**Comment by ejgallego** [comment]
@Zimmi48 I don't remember the context of my comment on the OCaml issue.

We could try 128MiB and see if that works, that's still hacky but IMHO would save us work for what is really a quite corner case in our setup.

**Comment by SkySkimmer** [comment]
I don't think fiat_crypto_ocaml runs anything that the minimizer understands.

It's also pretty useless and takes a while so we could just remove it.

**Comment by Zimmi48** [comment]
OK, in any case, the current version of the PR works.

**Comment by SkySkimmer** [comment]
@coqbot run light ci

**Comment by Zimmi48** [comment]
Light CI pipeline passed as well (whatever that was for).

**Comment by SkySkimmer** [comment]
@coqbot merge now
